archivist/archivistpublic.py

```diff
@@ -0,0 +1,428 @@
+
+   .. code-block:: python
+
+      # Initialize connection to Archivist
+      public = :Public(
+          max_time=300.0,
+      )
+
+    The public variable now has additional endpoints assets,events.
+
+"""
+
+from logging import getLogger
+from collections import deque
+from copy import deepcopy
+from typing import BinaryIO, Dict, List, Optional
+
+import requests
+from requests.models import Response
+
+
+from .confirmer import MAX_TIME
+from .constants import (
+    HEADERS_REQUEST_TOTAL_COUNT,
+    HEADERS_TOTAL_COUNT,
+)
+from .dictmerge import _deepmerge, _dotdict
+from .errors import (
+    _parse_response,
+    ArchivistBadFieldError,
+    ArchivistDuplicateError,
+    ArchivistHeaderError,
+    ArchivistNotFoundError,
+)
+from .headers import _headers_get
+from .retry429 import retry_429
+
+from .assets import _AssetsPublic
+from .events import _EventsPublic
+from .assetattachments import _AssetAttachmentsClient
+
+LOGGER = getLogger(__name__)
+
+
+class ArchivistPublic:  # pylint: disable=too-many-instance-attributes
+    """Base class for public Archivist endpoints.
+
+    This class manages the connection to an Archivist instance and provides
+    basic methods that represent the underlying REST interface.
+
+    Args:
+        verify: if True the certificate is verified
+        max_time (float): maximum time in seconds to wait for confirmation
+
+    """
+
+    # also change the type hints in __init__ below
+    CLIENTS = {
+        "assets": _AssetsPublic,
+        "events": _EventsPublic,
+        "assetattachments": _AssetAttachmentsClient,
+    }
+
+    RING_BUFFER_MAX_LEN = 10
+
+    def __init__(
+        self,
+        *,
+        fixtures: Optional[Dict] = None,
+        verify: bool = True,
+        max_time: float = MAX_TIME,
+    ):
+
+        self._verify = verify
+        self._response_ring_buffer = deque(maxlen=self.RING_BUFFER_MAX_LEN)
+        self._session = None
+        self._max_time = max_time
+        self._fixtures = fixtures or {}
+
+        # Type hints for IDE autocomplete, keep in sync with CLIENTS map above
+        self.assets: _AssetsPublic
+        self.events: _EventsPublic
+        self.assetattachments: _AssetAttachmentsClient
+
+    def __str__(self) -> str:
+        return "ArchivistPublic()"
+
+    def __getattr__(self, value: str):
+        """Create endpoints on demand"""
+        client = self.CLIENTS.get(value)
+
+        if client is None:
+            raise AttributeError
+
+        c = client(self)
+        super().__setattr__(value, c)
+        return c
+
+    @property
+    def session(self) -> requests.Session:
+        """creates and returns session"""
+        if self._session is None:
+            self._session = requests.Session()
+        return self._session
+
+    def close(self):
+        """closes current session"""
+        if self._session is not None:
+            self._session.close()
+            self._session = None
+
+    @property
+    def public(self) -> str:
```

Comment:
Maybe when I see the sample code I'll be less worried about this, but as things stand I'm concerned about the semantics of this function. It seems to encourage code like "if arch.public" to test whether it's a public or restricted connection as well as being used for prefixing object identities.

Isn't that dual use a bit fragile? Are we certain that restricted assets (or some other as-yet unimagined subclass) will never have anything other than empty string here?

Reply:
I am looking at this now...

Reply:
well this uses python's concept of truthiness - if the public()  is an empty string then it is not public...